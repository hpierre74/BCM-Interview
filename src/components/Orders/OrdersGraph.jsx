import React from "react";
import {
  LineChart,
  Line,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
  Brush
} from "recharts";
import moment from "moment";

const OrdersGraph = props => {

  // Unnecessary loop to improve user dates readability
  // Workaround / additional instructions about date displaying needed
  const groupDates = () => {
    const newData = [];
    Object.values(props.data).map((order, index) => {
      order.date = moment(order.datetime).format("DD/MM/YYYY");
      return newData.push(order);
    });
    return newData;
  };

  return (
    <div>
      <h1> {props.title} </h1>
      <LineChart
        width={window.innerWidth / 2.5}
        height={300}
        data={groupDates()}
        margin={{ top: 5, right: 30, left: 20, bottom: 5 }}
      >
        <XAxis dataKey="date" />
        <YAxis />
        <CartesianGrid strokeDasharray="3 3" />
        <Tooltip />
        <Legend />
        <Brush dataKey="date" height={30} stroke={props.primary} />
        <Line
          type="monotone"
          dataKey="price"
          stroke={props.primary}
          activeDot={{ r: 8 }}
        />
        <Line type="monotone" dataKey="datetime" stroke={props.secondary} />
      </LineChart>
    </div>
  );
};

export default OrdersGraph;
