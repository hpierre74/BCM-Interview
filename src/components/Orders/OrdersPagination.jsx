import React from "react";
import { Pagination, PaginationItem, PaginationLink } from "reactstrap";

const OrdersPagination = props => {
  const { orders, page, paginate, isVisible, displayPagination } = props;

  const renderPaginations = () => {
    return Object.values(orders).map((order, index) => {
      return (
        <PaginationItem key={index}>
          <PaginationLink onClick={e => paginate(e, index)} value={index}>
            {index}
          </PaginationLink>
        </PaginationItem>
      );
    });
  };
  const styles = {
    justifyContent: "flex-end",
    maxWidth: "100%",
    display: "flex",
    flexFlow: "row wrap"
  };
  return (
    <Pagination style={styles}>
      <PaginationItem>
        <PaginationLink onClick={e => paginate(e, page - 1)} previous />
      </PaginationItem>
      {isVisible ? (
        renderPaginations()
      ) : (
        <PaginationItem>
          <PaginationLink> {page} </PaginationLink>
        </PaginationItem>
      )}
      <PaginationItem>
        <PaginationLink onClick={e => displayPagination(e)}>
          {" "}
          ...{" "}
        </PaginationLink>
      </PaginationItem>
      <PaginationItem>
        <PaginationLink next onClick={e => paginate(e, page + 1)} />
      </PaginationItem>
    </Pagination>
  );
};
export default OrdersPagination;
