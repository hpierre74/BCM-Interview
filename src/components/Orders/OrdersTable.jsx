import React from "react";
import { Table } from "reactstrap";
import moment from "moment";

const OrdersTable = props => {
  const { orders } = props;
  const renderOrders = () => {
    return Object.values(orders).map((order, index) => {
      return (
        <tr key={index}>
          <td> {moment(order.datetime).format("DD/MM/YYYY HH:mm:ss")}</td>
          <td> {order.type} </td>
          <td> {order.volume} </td>
          <td> {order.price + " " + order.currency} </td>
        </tr>
      );
    });
  };
  return (
    <div>
      <Table striped className="OrdersLayout">
        <thead>
          <tr className="OrdersLayout-header">
            <th>Date</th>
            <th>Order</th>
            <th>Volume</th>
            <th>Price</th>
          </tr>
        </thead>
        <tbody>{renderOrders()}</tbody>
      </Table>
    </div>
  );
};

export default OrdersTable;
