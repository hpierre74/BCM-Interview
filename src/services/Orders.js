import axios from "axios";

const orderInstance = axios.create({
  baseURL: "http://localhost:3001/orders/"
});

class Orders {
  static getOrders(type, cb, page) {
    return page
      ? orderInstance
          .get(type + "?page=" + page)
          .then(res => {
            return page
              ? cb(res.data.orders, type, true)
              : cb(res.data.orders, type, false);
          })
          .catch(err => console.log(err))
      : orderInstance
          .get(type)
          .then(res => {
            return cb(res.data.orders, type);
          })
          .catch(err => console.log(err));
  }
}
export default Orders;
