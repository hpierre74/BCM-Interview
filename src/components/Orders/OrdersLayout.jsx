import React, { Component } from "react";
import { Container, Row, Col } from "reactstrap";
import axios from "axios";
import OrdersTable from "./OrdersTable";
import OrdersPagination from "./OrdersPagination";
import OrdersGraph from "./OrdersGraph";

class OrdersLayout extends Component {
  constructor(context, props) {
    super(context, props);
    this.state = {
      page: 0,
      allOrders: [],
      isPaginationVisible: false,
      chunkedOrders: [],
      sellOrders: [],
      buyOrders: []
    };
    this.req = axios.create({
      baseURL: "http://localhost:3001/orders/"
    });
    //I skipped pagination requests as it wasn't technically hard to implement
    //I would normally use the pagination or set data in session/cache/cookies
    //to avoid multiple requests
    //this.paging = "?page=" + this.state.page;
  }

  componentWillMount() {
    this.setOrders();
  }

  setOrders() {
    this.getOrders("ALL", "allOrders");
    this.getOrders("SELL", "sellOrders");
    this.getOrders("BUY", "buyOrders");
  }

  getOrders(type, state) {
    this.req
      .get(type)
      .then(res => {
        return this.setState(
          {
            [state]: res.data.orders
          },
          () => {
            return state === "allOrders"
              ? this.chunkOrders(res.data.orders)
              : "";
          }
        );
      })
      .catch(err => console.log(err));
  }

  //Set the current Table page value
  paginate(e, page) {
    e.preventDefault();
    this.setState({ page: page });
  }

  //toggle Pagination links visibility
  displayPagination(e) {
    e.preventDefault();
    this.setState({ isPaginationVisible: !this.state.isPaginationVisible });
  }

  //Change the array to multiple ones with length relying on given chunk_size param
  chunkArray(myArray, chunk_size) {
    let results = [];
    while (myArray.length) {
      results.push(myArray.splice(0, chunk_size));
    }
    return results;
  }

  chunkOrders(orders) {
    return this.setState({ chunkedOrders: this.chunkArray(orders, 10) });
  }

  render() {
    return (
      <Container fluid>
        <Row>
          <Col style={{ backgroundColor: "#434449" }} sm="2" md="2" xs="2">
            Orders...
          </Col>
          <Col>
            <Row>
              <Col>
                <OrdersTable
                  orders={
                    this.state.chunkedOrders[this.state.page]
                      ? this.state.chunkedOrders[this.state.page]
                      : []
                  }
                />
                <OrdersPagination
                  isVisible={this.state.isPaginationVisible}
                  displayPagination={this.displayPagination.bind(this)}
                  page={this.state.page}
                  paginate={this.paginate.bind(this)}
                  orders={this.state.chunkedOrders}
                />
              </Col>
            </Row>
            <Row>
              <Col style={{ display: "flex", flexFlow: "row wrap" }}>
                <OrdersGraph
                  title="Selling Orders"
                  primary="#FF5733"
                  secondary="#FF8A33"
                  data={this.state.sellOrders}
                />
                <OrdersGraph
                  title="Buying Orders"
                  primary="#8884d8"
                  secondary="#82ca9d"
                  data={this.state.buyOrders}
                />
              </Col>
            </Row>
          </Col>
        </Row>
      </Container>
    );
  }
}

export default OrdersLayout;
