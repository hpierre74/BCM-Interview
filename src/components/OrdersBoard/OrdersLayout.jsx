import React, { Component } from "react";
import { Container, Row, Col } from "reactstrap";
import OrdersTable from "./OrdersTable";
import OrdersPagination from "./OrdersPagination";
import OrdersGraph from "./OrdersGraph";
import OrdersNav from './OrdersNav';
import Orders from "../../services/Orders";

import './OrdersBoard.css';


class OrdersLayout extends Component {
  constructor(props) {
    super(props);
    this.state = {
      BUY: [],
      SELL: [],
      current:[],
      page:0
    };
    this.saveOrders = this.saveOrders.bind(this);
  }

  componentWillMount() {
    this.setOrders();
  }

  saveOrders(orders, type, isPage) {
    return isPage
      ? this.setState({
          current: orders
        })
      : this.setState({
          [type]: orders
        });
  }

  setOrders() {
    Orders.getOrders("ALL", this.saveOrders, this.state.page.toString());
    Orders.getOrders("BUY", this.saveOrders);
    Orders.getOrders("SELL", this.saveOrders);
  }  

  
  //Set the current Table page value
  paginate(e, page) {
    e.preventDefault();
    page = page < 0 ? page = 0 : page; // prevent overload
    Orders.getOrders("ALL", this.saveOrders, page.toString());
    this.setState({ page: page });
  }

  render() {
    return (
      <Container className='OrdersLayout' fluid>
        <Row>
          <Col style={{ backgroundColor: "#434449" }} sm="2" md="2" xs="2">
            <OrdersNav />
          </Col>
          <Col  sm="10" md="10" xs="10">
            <Row>
              <Col>
                <h2>Orders Board</h2>
                <OrdersTable
                  orders={this.state.current}
                />
                <OrdersPagination
                  page={this.state.page}
                  paginate={this.paginate.bind(this)}
                />
              </Col>
            </Row>
            <Row>
              <Col style={{ display: "flex", flexFlow: "row wrap", justifyContent: 'center' }}>
                <OrdersGraph
                  title="Selling Orders"
                  primary="#FF5733"
                  secondary="#FF8A33"
                  data={this.state.SELL}
                />
                <OrdersGraph
                  title="Buying Orders"
                  primary="#8884d8"
                  secondary="#82ca9d"
                  data={this.state.BUY}
                />
              </Col>
            </Row>
          </Col>
        </Row>
      </Container>
    );
  }
}

export default OrdersLayout;
