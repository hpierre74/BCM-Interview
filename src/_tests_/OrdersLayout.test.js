import OrdersLayout from '../components/OrdersBoard/OrdersLayout';
import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Orders from '../services/orders';


describe('OrdersLayout component renders the children correctly', () => {
    it('renders correctly', () => {  
        const rendered = renderer.create(
        <OrdersLayout />
      );
      expect(rendered.toJSON()).toMatchSnapshot();
    });
  });
