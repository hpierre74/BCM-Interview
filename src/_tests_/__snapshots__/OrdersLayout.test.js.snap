// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdersLayout component renders the children correctly renders correctly 1`] = `
<div
  className="OrdersLayout container-fluid"
>
  <div
    className="row"
  >
    <div
      className="col-2 col-sm-2 col-md-2"
      style={
        Object {
          "backgroundColor": "#434449",
        }
      }
    >
      <ul
        className="nav flex-column"
      >
        <li
          className="nav-item"
        >
          <a
            className="nav-link"
            href="#"
            onClick={[Function]}
          >
             
            Wind power
             
          </a>
        </li>
        <li
          className="nav-item"
        >
          <a
            className="nav-link"
            href="#"
            onClick={[Function]}
          >
             
            Biomass
             
          </a>
        </li>
        <li
          className="nav-item"
        >
          <a
            className="nav-link"
            href="#"
            onClick={[Function]}
          >
             
            Solar
             
          </a>
        </li>
        <li
          className="nav-item"
        >
          <a
            className="nav-link"
            href="#"
            onClick={[Function]}
          >
             
            Gas
             
          </a>
        </li>
      </ul>
    </div>
    <div
      className="col-10 col-sm-10 col-md-10"
    >
      <div
        className="row"
      >
        <div
          className="col"
        >
          <h2>
            Orders Board
          </h2>
          <div>
            <table
              className="OrdersTable paper table table-striped"
            >
              <thead>
                <tr
                  className="OrdersTable-header"
                >
                  <th>
                    Date
                  </th>
                  <th>
                    Order
                  </th>
                  <th>
                    Volume
                  </th>
                  <th>
                    Price
                  </th>
                </tr>
              </thead>
              <tbody />
            </table>
          </div>
          <ul
            className="pagination"
            style={
              Object {
                "display": "flex",
                "flexFlow": "row wrap",
                "justifyContent": "flex-end",
                "maxWidth": "100%",
              }
            }
          >
            <li
              className="page-item"
            >
              <a
                aria-label="Previous"
                className="page-link"
                onClick={[Function]}
              >
                <span
                  aria-hidden="true"
                >
                  «
                </span>
                <span
                  className="sr-only"
                >
                  Previous
                </span>
              </a>
            </li>
            <li
              className="page-item"
            >
              <a
                aria-label={undefined}
                className="page-link"
              >
                 
                0
                 
              </a>
            </li>
            <li
              className="page-item"
            >
              <a
                aria-label="Next"
                className="page-link"
                onClick={[Function]}
              >
                <span
                  aria-hidden="true"
                >
                  »
                </span>
                <span
                  className="sr-only"
                >
                  Next
                </span>
              </a>
            </li>
            <input
              className="form-control"
              id="pickPage"
              name="pickPage"
              onChange={[Function]}
              placeholder="go to"
              style={
                Object {
                  "width": "100px",
                }
              }
              type="number"
            />
          </ul>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col"
          style={
            Object {
              "display": "flex",
              "flexFlow": "row wrap",
              "justifyContent": "center",
            }
          }
        >
          <div
            style={
              Object {
                "justifyContent": "center",
              }
            }
          >
            <h4>
               
              Selling Orders
               
            </h4>
            <div
              className="recharts-wrapper"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseMove={[Function]}
              onTouchMove={[Function]}
              style={
                Object {
                  "cursor": "default",
                  "height": 300,
                  "margin": "5px auto",
                  "position": "relative",
                  "width": 700,
                }
              }
            >
              <svg
                className="recharts-surface"
                height={300}
                style={undefined}
                version="1.1"
                viewBox="0 0 700 300"
                width={700}
              >
                <g
                  className="recharts-cartesian-grid"
                >
                  <g
                    className="recharts-cartesian-grid-horizontal"
                  >
                    <line
                      fill="none"
                      height={230}
                      stroke="#ccc"
                      strokeDasharray="3 3"
                      width={640}
                      x={60}
                      x1={60}
                      x2={700}
                      y={5}
                      y1={5}
                      y2={5}
                    />
                    <line
                      fill="none"
                      height={230}
                      stroke="#ccc"
                      strokeDasharray="3 3"
                      width={640}
                      x={60}
                      x1={60}
                      x2={700}
                      y={5}
                      y1={235}
                      y2={235}
                    />
                  </g>
                  <g
                    className="recharts-cartesian-grid-vertical"
                  >
                    <line
                      fill="none"
                      height={230}
                      stroke="#ccc"
                      strokeDasharray="3 3"
                      width={640}
                      x={60}
                      x1={60}
                      x2={60}
                      y={5}
                      y1={5}
                      y2={235}
                    />
                    <line
                      fill="none"
                      height={230}
                      stroke="#ccc"
                      strokeDasharray="3 3"
                      width={640}
                      x={60}
                      x1={700}
                      x2={700}
                      y={5}
                      y1={5}
                      y2={235}
                    />
                  </g>
                </g>
              </svg>
              <div
                className="recharts-legend-wrapper"
                style={
                  Object {
                    "bottom": 5,
                    "height": "auto",
                    "left": 0,
                    "position": "absolute",
                    "width": 700,
                  }
                }
              >
                <ul
                  className="recharts-default-legend"
                  style={
                    Object {
                      "margin": 0,
                      "padding": 0,
                      "textAlign": "center",
                    }
                  }
                >
                  <li
                    className="recharts-legend-item legend-item-0"
                    style={
                      Object {
                        "display": "inline-block",
                        "marginRight": 10,
                      }
                    }
                  >
                    <svg
                      className="recharts-surface"
                      height={14}
                      style={
                        Object {
                          "display": "inline-block",
                          "marginRight": 4,
                          "verticalAlign": "middle",
                        }
                      }
                      version="1.1"
                      viewBox="0 0 32 32"
                      width={14}
                    >
                      <path
                        className="recharts-legend-icon"
                        d="M0,16h10.666666666666666
                                  A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
                                  H32M21.333333333333332,16
                                  A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
                        fill="none"
                        stroke="#FF5733"
                        strokeWidth={4}
                      />
                    </svg>
                    <span
                      className="recharts-legend-item-text"
                    >
                      price
                    </span>
                  </li>
                </ul>
              </div>
              <div
                className="recharts-tooltip-wrapper"
                style={
                  Object {
                    "MozTransform": "translate(undefinedpx, undefinedpx)",
                    "OTransform": "translate(undefinedpx, undefinedpx)",
                    "WebkitTransform": "translate(undefinedpx, undefinedpx)",
                    "msTransform": "translate(undefinedpx, undefinedpx)",
                    "pointerEvents": "none",
                    "position": "absolute",
                    "top": 0,
                    "transform": "translate(undefinedpx, undefinedpx)",
                    "visibility": "hidden",
                  }
                }
              >
                <div
                  className="recharts-default-tooltip"
                  style={
                    Object {
                      "backgroundColor": "#fff",
                      "border": "1px solid #ccc",
                      "margin": 0,
                      "padding": 10,
                      "whiteSpace": "nowrap",
                    }
                  }
                >
                  <p
                    className="recharts-tooltip-label"
                    style={
                      Object {
                        "margin": 0,
                      }
                    }
                  >
                    
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            style={
              Object {
                "justifyContent": "center",
              }
            }
          >
            <h4>
               
              Buying Orders
               
            </h4>
            <div
              className="recharts-wrapper"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseMove={[Function]}
              onTouchMove={[Function]}
              style={
                Object {
                  "cursor": "default",
                  "height": 300,
                  "margin": "5px auto",
                  "position": "relative",
                  "width": 700,
                }
              }
            >
              <svg
                className="recharts-surface"
                height={300}
                style={undefined}
                version="1.1"
                viewBox="0 0 700 300"
                width={700}
              >
                <g
                  className="recharts-cartesian-grid"
                >
                  <g
                    className="recharts-cartesian-grid-horizontal"
                  >
                    <line
                      fill="none"
                      height={230}
                      stroke="#ccc"
                      strokeDasharray="3 3"
                      width={640}
                      x={60}
                      x1={60}
                      x2={700}
                      y={5}
                      y1={5}
                      y2={5}
                    />
                    <line
                      fill="none"
                      height={230}
                      stroke="#ccc"
                      strokeDasharray="3 3"
                      width={640}
                      x={60}
                      x1={60}
                      x2={700}
                      y={5}
                      y1={235}
                      y2={235}
                    />
                  </g>
                  <g
                    className="recharts-cartesian-grid-vertical"
                  >
                    <line
                      fill="none"
                      height={230}
                      stroke="#ccc"
                      strokeDasharray="3 3"
                      width={640}
                      x={60}
                      x1={60}
                      x2={60}
                      y={5}
                      y1={5}
                      y2={235}
                    />
                    <line
                      fill="none"
                      height={230}
                      stroke="#ccc"
                      strokeDasharray="3 3"
                      width={640}
                      x={60}
                      x1={700}
                      x2={700}
                      y={5}
                      y1={5}
                      y2={235}
                    />
                  </g>
                </g>
              </svg>
              <div
                className="recharts-legend-wrapper"
                style={
                  Object {
                    "bottom": 5,
                    "height": "auto",
                    "left": 0,
                    "position": "absolute",
                    "width": 700,
                  }
                }
              >
                <ul
                  className="recharts-default-legend"
                  style={
                    Object {
                      "margin": 0,
                      "padding": 0,
                      "textAlign": "center",
                    }
                  }
                >
                  <li
                    className="recharts-legend-item legend-item-0"
                    style={
                      Object {
                        "display": "inline-block",
                        "marginRight": 10,
                      }
                    }
                  >
                    <svg
                      className="recharts-surface"
                      height={14}
                      style={
                        Object {
                          "display": "inline-block",
                          "marginRight": 4,
                          "verticalAlign": "middle",
                        }
                      }
                      version="1.1"
                      viewBox="0 0 32 32"
                      width={14}
                    >
                      <path
                        className="recharts-legend-icon"
                        d="M0,16h10.666666666666666
                                  A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
                                  H32M21.333333333333332,16
                                  A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16"
                        fill="none"
                        stroke="#8884d8"
                        strokeWidth={4}
                      />
                    </svg>
                    <span
                      className="recharts-legend-item-text"
                    >
                      price
                    </span>
                  </li>
                </ul>
              </div>
              <div
                className="recharts-tooltip-wrapper"
                style={
                  Object {
                    "MozTransform": "translate(undefinedpx, undefinedpx)",
                    "OTransform": "translate(undefinedpx, undefinedpx)",
                    "WebkitTransform": "translate(undefinedpx, undefinedpx)",
                    "msTransform": "translate(undefinedpx, undefinedpx)",
                    "pointerEvents": "none",
                    "position": "absolute",
                    "top": 0,
                    "transform": "translate(undefinedpx, undefinedpx)",
                    "visibility": "hidden",
                  }
                }
              >
                <div
                  className="recharts-default-tooltip"
                  style={
                    Object {
                      "backgroundColor": "#fff",
                      "border": "1px solid #ccc",
                      "margin": 0,
                      "padding": 10,
                      "whiteSpace": "nowrap",
                    }
                  }
                >
                  <p
                    className="recharts-tooltip-label"
                    style={
                      Object {
                        "margin": 0,
                      }
                    }
                  >
                    
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
